Add private key generation

AndrzejKurek opened the pull request:
This PR is a prerequisite for https://github.com/ARMmbed/mbed-os-example-atecc608a/pull/4.

Commit:
Add private key generation

gilles-peskine-arm commented on atecc608a_se.c:
```diff
@@ -211,6 +211,53 @@ static psa_status_t atecc608a_import_public_key(psa_key_slot_number_t key_slot,
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (bits != PSA_BYTES_TO_BITS(ATCA_PRIV_KEY_SIZE))
```
The core guarantees that `bits` is consistent with `type`, so this check is redundant. But it's ok to have it for robustness.

Patater replied:
Should this be an assert to make sure the core didn't pass us junk?

AndrzejKurek replied:
Since I wasn't sure about the constraints that the core will have, I opted for not using asserts and just returning errors instead.

Patater replied:
We don't have any mechanism for asserting right now. I think we should consider adding it, but outside the scope of this PR.

Patater commented on atecc608a_se.c:
```diff
@@ -211,6 +211,53 @@ static psa_status_t atecc608a_import_public_key(psa_key_slot_number_t key_slot,
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* The hardware has slots 0-15 */
+    if (key_slot > 15)
```
Will `psa_key_slot_number_t` always be unsigned?

Isn't `0` an invalid slot? or has the translation from user-facing slot numbers to hardware slot numbers happened before the driver call?

I would say it makes more sense for the driver itself to perform the translation from user-facing slot numbers to hardware slot numbers, since how to do the translation may vary based on the particular hardware. In this case, the driver should be disallowing 0 unless this check has already been performed before the driver is called. It's programmer error on Mbed Crypto's part if 0 is passed as a user-facing slot number to a driver. (So maybe we just need an assert, not a check.)

AndrzejKurek replied:
I assumed that psa_key_slot_number_t passed to this function is already translated to the hardware slot number, as I believe that that was the plan a couple of weeks ago when I looked into the key slot code update. If so, then 0 is a perfectly fine number for our hardware, as are numbers up to 15, inclusive. There's no reason from this hardware perspective for it to be signed, but as for the driver - I don't have an opinion.

I can imagine that there are cryptochips that don't have slot numbers, but an arbitrary identifier of an opaque sort, and then we'd need to have the number translation here, to maintain a consistent API (or just have a void* parameter).

As for the assert vs error - I assumed that there may be cases where the user requests an access to slots incrementally up until an error is returned, and generally, returning an error lets the user to handle the situation in any manner he wants.

Patater replied:
I don't think we can assume translation from user-facing slot numbers to hardware slot numbers has already occurred at this point. Akin to how the pubkey is in user-facing format on output, the inputs should also be in user-facing format. The driver should control how it does key format translation as well as slot number translation.

AndrzejKurek replied:
But this code is not called directly by the user, it will be called through the crypto SE driver API (or rather, PSA will call it), so I assumed that this number will be hidden from the user. 
If what you write is true, how would slot allocation work? I thought that the user registers or allocates a slot (passing the requested hardware number, since he knows which slot exactly he wants to use) and receives an additional, user-friendly ID. Then, when requesting an operation, PSA translates it before each call to the driver itself.
If that's not the case, each operation would need additional number translation and separate management of slot mapping, and that seems to be a bit of an overkill.

Patater replied:
OK, will resolve as part of https://github.com/ARMmbed/psa-crypto/issues/110#issuecomment-498738802 I believe.

gilles-peskine-arm replied:
The whole concept of slot numbers is still under discussion (https://github.com/ARMmbed/psa-crypto/issues/110). If they exist at all, it'll be some unsigned type, but where 0 is an ordinary value, because typical SEs number their slots from 0. Note that the slot number is not related to the key id (unless it is, but in that case the key id would have to be in a certain small range with a translation `key_id = slot_numer + OFFSET`).

> I would say it makes more sense for the driver itself to perform the translation from user-facing slot numbers to hardware slot numbers

Yes, but… see https://github.com/ARMmbed/psa-crypto/issues/110. The driver lacks storage to remember which slots are occupied, and it's difficult to express the logic for deciding which slot numbers are suitable for which keys.

> I can imagine that there are cryptochips that don't have slot numbers

PKCS#11 doesn't. You can set a label for the key, but you might as well use the key id, there's no need to invent yet another number. At least assuming that there isn't a stringent (&lt;8B) size limitation on the label.

Patater commented on atecc608a_se.c:
```diff
@@ -211,6 +211,53 @@ static psa_status_t atecc608a_import_public_key(psa_key_slot_number_t key_slot,
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (p_pubkey_out != NULL && pubkey_out_size < ATCA_PUB_KEY_SIZE)
```
Is `pubkey_out_size` in user-facing pubkey format or internal hardware-facing pubkey format? Don't we need some sort of size translation here?

I assume `pubkey_out_size` is in user-facing pubkey format, which is different from the hardware format. Maybe the user-facing pubkey format requires a `0x4` on the front, and the size given includes that extra byte.

If the user-facing format and the hardware format do indeed occur, did we test that if the driver is called with a buffer that is large enough for the hardware format, but not large enough for the user-facing format, we fail as expected?

AndrzejKurek replied:
That's an oversight on my part - this code lacks the format translation that the public key exporting code has. I'll add that, and as for the size testing - I'll add various tests as the first step towards interactive app refactoring, in a separate PR.

AndrzejKurek replied:
Updated.

Commit:
Change the format of a public key exported when generating a private key
Add a '0x04' prefix before the x and y parts of the key to be compliant
with the PSA format.

Patater commented on atecc608a_se.c:
```diff
@@ -240,17 +247,22 @@ static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
+
+    /* atcab_genkey returns concatenated x and y values, and the desired
+     * format is 0x04 + x + y. Start at &p_pubkey_out[1] and add a 0x04
+     * at p_pubkey_out[0]. */
```
Would you agree that its worth making a static helper functions to do format translation? It looks like any function dealing with pubkeys will need shared format translation code. I'd imagine we need one for driver_to_user and user_to_driver.

AndrzejKurek replied:
We can either leave it as is, calling the cryptoauthlib with `&p_pubkey_out[1]` and adding a `0x04`, or increase the complexity by calling it at `&p_pubkey_out[0]`, then calling an utility function that would `memmove` it and add `0x04` at the beginning.
Since all the translation does is essentially adding / stripping off `0x04`, I would rather leave it as it is right now.

Patater replied:
We shouldn't need to use `memmove()`.

Maybe something like this would help us avoid having to repeat ourselves.

```C
/* The driver works with pubkeys as concatenated x and y values, and the PSA
 * format for pubkeys is 0x04 + x + y. Always use a pubkey buffer in PSA
 * format, with enough space for the PSA format. To translate this buffer for
 * use with cryptoauthlib, use pubkey_for_driver(). To ensure the buffer is in
 * valid PSA format after cryptoauthlib operations, call pubkey_for_psa(). */
static uint8_t *pubkey_for_driver(uint8_t *data)
{
    return &data[1];
}

static void pubkey_for_psa(uint8_t *data)
{
    data[0] = 0x4;
}

/* ... */
{
    ASSERT_SUCCESS(atcab_genkey(key_id, pubkey_for_driver(p_pubkey_out)));
    pubkey_for_psa(p_pubkey_out);
}
/* ... */
```

gilles-peskine-arm replied:
I think the code is less readable with these trivial helpers, since you have to read more code to figure out whether the buffer is processed correctly.

Patater replied:
You have to read less code in total, when reviewing the entire driver, because only one function contains this logic. You are less likely to make mistakes when handling the format translation, since you have known-good functions to use instead of copy-paste; and not using the helpers is obvious in code review.

gilles-peskine-arm commented on atecc608a_se.c:
```diff
@@ -240,17 +247,29 @@ static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
     }
 
-    if (p_pubkey_out != NULL && pubkey_out_size < ATCA_PUB_KEY_SIZE)
+    if (p_pubkey_out != NULL && pubkey_out_size < key_data_len)
```
Actually you're going to write `1 + ATCA_PUB_KEY_SIZE` bytes, not some amount calculated from PSA macros. So the correct test here would be `pubkey_out_size < 1 + ATCA_PUB_KEY_SIZE`.

AndrzejKurek replied:
I believe that that's the way I wrote it in my first approach. Do you agree for that, @Patater?

Patater replied:
For consistency, when writing to a buffer used by the PSA Crypto core, I'd recommend using sizes related to the PSA Crypto core formats. When writing to a buffer used by the driver, I'd recommend using driver sizes. This should help avoid using incorrect sizes. We also don't need to repeat the knowledge that `1 + ATCA_PUB_KEY_SIZE` is how to convert the size everywhere.

This is a stylistic choice, however. I'm happy with either style if we assert that we've written the correct number of bytes for the PSA format in a robust enough way that if the PSA format changes or the underlying driver format changes, we realize in an obvious way we need to update the format translation code.

gilles-peskine-arm replied:
Here, the core guarantees that the buffer is at least `PSA_KEY_EXPORT_MAX_SIZE(PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_CURVE_SECP256R1), 256)` bytes long (by the specification of the export functions), and that this is ≥65 (by the specification of the size macro). The check that's useful in the driver is that the buffer size is at least the number of bytes that the driver actually writes, and the driver writes 1 byte at offset 0 and `ATCA_PUB_KEY_SIZE` bytes at offset 1.

gilles-peskine-arm commented on atecc608a_se.c:
```diff
@@ -240,17 +247,29 @@ static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
     if (p_pubkey_length != NULL)
     {
-        *p_pubkey_length = ATCA_PUB_KEY_SIZE;
+        *p_pubkey_length = key_data_len;
```
Again, you copied `1 + ATCA_PUB_KEY_SIZE` bytes.

gilles-peskine-arm reviewed (changes requested):
This is functionally correct but not exactly right.

Commit:
Change calculation of required buffer size

Commit:
Move public key format translations to dedicated functions

AndrzejKurek commented:
Adjusted according to comments. @gilles-peskine-arm, @Patater - please re-review.

Patater commented:
What testing has been done with this?

AndrzejKurek commented:
@Patater:


    /* Valid values */
    psa_status_t status;
    static uint8_t pubkey[pubkey_size];
    size_t pubkey_len = 0;

    /* Invalid values */
    const uint16_t bad_key_id = 16;
    const psa_key_type_t bad_key_type = PSA_KEY_TYPE_RSA_PUBLIC_KEY;
    const size_t bad_key_bits = 5;
    const size_t bad_buffer_size = 64;

    /* Passing an invalid key slot should fail. */
    ASSERT_STATUS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         bad_key_id, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, pubkey, pubkey_size, &pubkey_len),
                  PSA_ERROR_INVALID_ARGUMENT,
                  PSA_ERROR_HARDWARE_FAILURE);

    /* Passing an invalid key type should fail. */
    ASSERT_STATUS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, bad_key_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, pubkey, pubkey_size,
                         &pubkey_len),
                  PSA_ERROR_NOT_SUPPORTED,
                  PSA_ERROR_HARDWARE_FAILURE);

    /* Passing invalid key bits should fail. */
    ASSERT_STATUS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         bad_key_bits, NULL, 0, pubkey, pubkey_size,
                         &pubkey_len),
                  PSA_ERROR_NOT_SUPPORTED,
                  PSA_ERROR_HARDWARE_FAILURE);

    /* Passing an invalid size should fail. */
    ASSERT_STATUS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, pubkey, bad_buffer_size,
                         &pubkey_len),
                  PSA_ERROR_BUFFER_TOO_SMALL,
                  PSA_ERROR_HARDWARE_FAILURE);

    /* Passing a NULL public key buffer should work, regardless of its size. */
    ASSERT_SUCCESS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, NULL, pubkey_size, &pubkey_len));

    /* Passing a NULL pubkey_len should work, even when exporting a public key. */
    ASSERT_SUCCESS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, pubkey, pubkey_size, NULL));

    /* Test that a public key received during a private key generation
     * can be imported. */
    ASSERT_SUCCESS_PSA(atecc608a_drv_info.p_key_management->p_generate(
                         atecc608a_private_key_slot, keypair_type,
                         PSA_KEY_USAGE_SIGN | PSA_KEY_USAGE_VERIFY,
                         key_bits, NULL, 0, pubkey, pubkey_size, &pubkey_len));

    ASSERT_SUCCESS_PSA(atecc608a_drv_info.p_key_management->p_import(
                         atecc608a_public_key_slot,
                         atecc608a_drv_info.lifetime,
                         key_type, alg, PSA_KEY_USAGE_VERIFY, pubkey,
                         pubkey_len));`
